plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -0,0 +1,570 @@
+                    .append("-").append(incrementalsVersion).append(".hpi").toString();
+
+            urlString = new StringBuffer(JENKINS_INCREMENTALS_REPO_MIRROR).append("/").append(groupId).append("/").
+                    append(incrementalsVersionPath).toString();
+        } else {
+            String pathToPlugin = new StringBuffer(pluginName).append("/").append(pluginVersion).append("/").
+                    append(pluginName).append(".hpi").toString();
+            urlString = new StringBuffer(JENKINS_UC_DOWNLOAD).append("/plugins/").append(pathToPlugin).toString();
+        }
+
+        System.out.println("Downloading plugin: " + pluginName + " from url: " + urlString);
+
+        File pluginFile = new File(refDir + SEPARATOR + plugin.getArchiveFileName());
+
+        try {
+            downloadToFile(urlString, pluginFile);
+        } catch (IOException | URISyntaxException e) {
+            e.printStackTrace();
+            failedPlugins.add(plugin);
+            System.out.println("Failed to download from requested URL");
+            return false;
+        }
+
+        //check integrity of plugin file
+        try {
+            JarFile pluginJpi = new JarFile(pluginFile);
+        } catch (IOException e) {
+            e.printStackTrace();
+            failedPlugins.add(plugin);
+            System.out.println("Downloaded file is not a valid ZIP");
+            return false;
+        }
+        return true;
+    }
+
+
+    public void downloadToFile(String urlString, File pluginFile) throws IOException, URISyntaxException {
+        CloseableHttpClient httpclient = HttpClients.createDefault();
+        HttpClientContext context = HttpClientContext.create();
+        HttpGet httpget = new HttpGet(urlString);
+        CloseableHttpResponse response = httpclient.execute(httpget, context);
+
+        try {
+            HttpHost target = context.getTargetHost();
+            List<URI> redirectLocations = context.getRedirectLocations();
+            // Expected to be an absolute URI
+            URI location = URIUtils.resolve(httpget.getURI(), target, redirectLocations);
+            FileUtils.copyURLToFile(location.toURL(), pluginFile);
+        } finally {
+            response.close();
+        }
+    }
+
+
+    public String getJenkinsVersion() {
+        //java -jar $JENKINS_WAR --version
+        try {
+            JarFile jenkinsWar = new JarFile(jenkinsWarFile);
+            Manifest manifest = jenkinsWar.getManifest();
+            Attributes attributes = manifest.getMainAttributes();
+            return attributes.getValue("Jenkins-Version");
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+
+        return "";
+    }
+
+
+    public String getPluginVersion(File file) {
+        //this is also done in the existing plugin manager from core - should I do this a similar way to that instead?
```

Comment:
what do you mean here? getting the manifest should be the right way to get the downloaded version

Reply:
I mean that there are very similar methods in Jenkins core that parse the manifest in a different manner.  I tried to do this in a straightforward way but I didn't know if one way was better than another. 

https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/PluginManager.java#L841

https://github.com/jenkinsci/jenkins/blob/master/core/src/main/java/hudson/PluginManager.java#L1058

Reply:
@timja Closed this PR since it was becoming hard to review. If you think that there are better ways to parse the manifest, please let me know.